설정 화면의 기본적인 UI와 Coordinator Flow, 팝업뷰 구현

sanghyeok-kim opened the pull request:
### 🔖 관련 이슈
- #20 

<br> 

### ⚒️ 작업 내역

- [x] FeatureSetting에 필요한 모듈들 구현(Coordinator, Presentation, UseCase...)
- [x] 설정 화면 기본적인 UI 구현
- [x] LifePoop 앱을 위한 BaseViewController 구현
(LifePoopViewController - backgroundColor, NavigationBar 관련 기본 설정)
- [ ] 프로필 정보 수정 화면 구현 
- [ ] 로그인 정보, 자동 로그인 여부, 버전 정보에 대한 UseCase 로직 구현
- [x] 약관 문서 출력을 위한 txt 파일 추가 및 Bundle Accessor 구현 
- [x] 서비스 이용 약관 화면 구현
- [x] 개인정보 처리 방침 화면 구현
- [ ] 의견 보내기 화면 구현 
- [x] 로그아웃 팝업 뷰 구현
- [x] 탈퇴 화면 기본적인 UI 구현 
- [ ] 탈퇴 화면에 사용되는 UITextView 커스터마이징, DesignSystem으로 분리

<br>

### 💻 리뷰어 가이드

- 앱 실행 후, 홈 화면 우측 상단의 설정 버튼을 누르면 설정 화면이 나타납니다.

<br>

### 📝 부가 설명

#### LifePoopViewController 구현
UIViewController를 구현할 때 발생하는 viewDidLoad, layoutUI, configureUI에 대한 코드 중복을 줄이고,
앱 화면의 일관성을 위해 BaseViewController가 필요하다고 생각되어  
DesignSystem에 LifePoopViewController를 구현하여 상속받도록 했습니다.

제드도 UIViewController들을 LifePoopViewController 서브클래스로 리팩토링해 주시고,  
주석 컨벤션은 제가 구현한 코드를 참조해서 다시 작성해주시면 됩니다.

#### Text 파일 읽기
Utils 모듈의 Resources 디렉토리에 .txt 파일을 모아뒀고, BundleAccessor를 이용해서 읽어올 수 있습니다.

#### 기타 사항
팝업뷰의 경우 0.2초 fade in & out 애니메이션을 넣어 자연스럽게 나타나도록 했습니다.

프로필 정보 화면은 제드가 만들어주신 TextField 화면을 사용하여 닉네임 변경 로직을 구현할 예정입니다.

탈퇴하기 화면에 있는 TextView는 의견 보내기 화면에서도 공통으로 사용할 수 있어  
DesignSystem에 별도로 구현할 계획입니다.

<br>

커밋 내용이 많아 주요 커밋만 아래에 정리했습니다.

- [feat: SettingType, SettingModel 구현, 각 Type별 Model과 ViewModel 구현](https://github.com/LifePoop/LifePoop_iOS/pull/42/commits/d0e032a86cb0c2faf39f6c451ea67ff796710b71)

- [feat: SettingCoordinator에 대한 로직 구현](https://github.com/LifePoop/LifePoop_iOS/pull/42/commits/08376390d87ed634e75e99d19e09ab4285933b5c)

- [feat: SettingViewController, SettingViewModel 생성 및 구현](https://github.com/LifePoop/LifePoop_iOS/pull/42/commits/c683a0d200c18b98375c54bbf45063270a0a9c51)

<br> 

### 📑 첨부 자료

https://github.com/LifePoop/LifePoop_iOS/assets/57667738/e2a14f2c-1041-466b-8378-6f4b742beada

sanghyeok-kim committed:
#20 feat: FeatureModuleType에 Setting 추가 및 Layer 의존성 배열 설정

sanghyeok-kim committed:
#20 feat: FeatureSetting 모듈별 Project 파일 생성 및 의존성 설정

sanghyeok-kim committed:
#20 feat: FeatureSetting의 각 모듈별 Sources, Tests 디렉토리와 플레이스 홀더 파일 생성

sanghyeok-kim committed:
#20 feat: SettingFeature에 필요한 이미지 에셋 추가

sanghyeok-kim committed:
#20 feat: FeatureHomeCoordinator 모듈에 SettingCoordinator, SettingCoordinatorInterface 모듈 의존성 추가

sanghyeok-kim committed:
#20 feat: HomeCoordinator에 SettingCoordinatorFlow 추가

sanghyeok-kim committed:
#20 feat: BaseSettingViewCell, BaseSettingTapViewCell, BaseSettingTextTapViewCell 구현

sanghyeok-kim committed:
#20 feat: SettingTableViewCell 구현 (LoginType, Switch, Tap, Text, TextTap)

sanghyeok-kim committed:
#20 feat: SettingModel 생성 및 설정 화면에 필요한 Model 구현

- LoginType, Profile, AutoLogin, Version, PrivacyPolicy, TermsOfService, SendFeedback

sanghyeok-kim committed:
#20 feat: SettingType 열거형 생성 및 구현

sanghyeok-kim committed:
#20 feat: LoginType 열거형 생성 및 구현

sanghyeok-kim committed:
#20 chore: Utils 모듈에 Resources 디렉토리 추가

sanghyeok-kim committed:
#20 feat: Utils/Resources에 PrivacyPolicy, TermsOfService txt파일 추가

sanghyeok-kim committed:
#20 feat: txt 파일에 접근할 수 있는 BundleAccessor와 이를 위한 DocumentType, TextFile 열거형 구현

sanghyeok-kim committed:
#20 refactor: SceneDelegate에서 NavigationBarBackButton의 기본 appearance 설정

- 기본적으로 나타나는 BackButton의 backIndicatorImage를 제거

sanghyeok-kim committed:
#20 refactor: SceneDelegate에서 NavigationBar BackButtonItem의 기본 appearance 설정

- 기본적으로 나타나는 BackButtonItem의 backIndicatorImage를 제거

sanghyeok-kim committed:
#20 feat: 앱 기본 BackBarButtonItem인 LifePoopBackBarButtonItem 생성 및 구현

sanghyeok-kim committed:
#20 feat: 앱 기본 ViewController인 LifePoopViewController 생성 및 구현

sanghyeok-kim committed:
#20 feat: LifePoopAlertViewType, LifePoopAlertView 생성 및 구현

sanghyeok-kim committed:
#20 feat: 문항 선택시 사용하는 SelectableTextRadioButton 생성 및 구현

sanghyeok-kim committed:
#20 chore: SettingTableViewCell들의 주석 수정

sanghyeok-kim committed:
#20 feat: SettingTextTapTableViewCell에 Tap Gesture 추가

sanghyeok-kim committed:
#20 feat: SettingType, SettingModel 구현, 각 Type별 Model과 ViewModel 구현

sanghyeok-kim committed:
#20 feat: SettingTableView의 DataSource, Delegate 생성 및 구현

sanghyeok-kim committed:
#20 feat: SettingTableView의 HeaderView, FooterView, FooterViewModel 생성 및 구현

sanghyeok-kim committed:
#20 feat: 탈퇴 화면을 위한 WithdrawReason 열거형 생성 및 구현

sanghyeok-kim committed:
#20 feat: 탈퇴 화면을 위한 WithdrawViewController, ViewModel 생성 및 구현

sanghyeok-kim committed:
#20 feat: Feedback, Profile 화면을 위한 ViewController, ViewModel 생성

sanghyeok-kim committed:
#20 feat: 개인정보 처리 방침, 서비스 이용 약관 화면을 위한 DocumentViewController, ViewModel 생성 및 구현

sanghyeok-kim committed:
#20 remove: 불필요한 Project Placeholder 파일 제거

sanghyeok-kim committed:
#20 feat: SettingCoordinator에 대한 로직 구현

sanghyeok-kim committed:
#20 feat: SettingDIContainer 생성 및 구현

sanghyeok-kim committed:
#20 feat: SettingViewController, SettingViewModel 생성 및 구현

sanghyeok-kim committed:
#20 fix: DocumentViewController의 title이 "프로필 정보 수정"으로 나타나는 문제 수정

junu0516 commented on Projects/DesignSystem/Sources/Component/UIButton/SelectableTextRadioButton.swift:
```diff
@@ -0,0 +1,48 @@
+    
+    public let index: Int
+    private let title: String
+    private let selectedImage = ImageAsset.btnRadioSelected.original
+    private let deselectedImage = ImageAsset.btnRadioDeselected.original
```
ImageAsset 사용할 때는 항상 `.original` 사용하도록 통일할까요?
기준이 좀 헷갈려서 제가 구현한 부분 다시 보니까 .image , .original이 혼재되있어서 질문남겨요!

sanghyeok-kim replied:
`.original`은 이미지의 tintColor 옵션을 alwaysOriginal로 설정한 프로퍼티인데요,
간혹 `.image`를 사용했을 때 이미지가 원본과 달라지는 경우가 있어서(기본 tintColor 색으로 변경됨)
`.original`을 사용하는걸로 통일하면 될 것 같아요

|.image|.original|
|---|---|
|<img width="182" alt="CleanShot 2023-05-22 at 21 51 06@2x" src="https://github.com/LifePoop/LifePoop_iOS/assets/57667738/ea4a25e3-7ba9-4f61-ad6d-831f7c4d4f69">|<img width="182" alt="CleanShot 2023-05-22 at 21 51 30@2x" src="https://github.com/LifePoop/LifePoop_iOS/assets/57667738/490617e4-4aba-4732-a548-b44bb1425ea3">|

junu0516 replied:
확인 감사합니다. 아마 이전에 ImageAsset 만드실 때 설명해주셨을텐데.. 제가 놓쳤던 것 같아요😅
.original로 통일해놓을께요!

junu0516 commented on Projects/DesignSystem/Sources/Component/UITableViewCell/BaseSettingTapTableViewCell.swift:
```diff
@@ -0,0 +1,41 @@
+
+import SnapKit
+
+open class BaseSettingTapTableViewCell: BaseSettingTableViewCell {
```
좀 더 해당 셀이 어느 유형의 설정(옵션)인지가 드러나는 이름이면 좋을 것 같아요

junu0516 commented on Projects/DesignSystem/Sources/Component/UITableViewCell/BaseSettingTextTapTableViewCell.swift:
```diff
@@ -0,0 +1,30 @@
+
+import SnapKit
+
+open class BaseSettingTextTapTableViewCell: BaseSettingTapTableViewCell {
```
바로 위와 동일합니다. 
각각의 옵션 유형에 따라 셀을 분류하신 의도는 바로 이해가는데,
LoginType, Switch 정도는 각기 로그인 정보 표시 혹은 스위치 버튼을 포함하는  뷰 표시 등의 의도가 전달되는데
Tap, Text, TextTap의 3가지는 이름만 보고는 어떤 형태의 뷰가 될 지 조금 헷갈려서 좀 더 구체적인 의미가 드러나면 좋을 것 같아요!

코드 내부를 각각 보면 이해가지만, 아무래도 다양한 종류의 셀을 포함하는 테이블뷰라면 셀의 이름이 좀 더 외부에서 봤을 때 의미가 직관적이면 좋지 않을까 생각했어요

sanghyeok-kim replied:
바로 위 코멘트에 대한 댓글도 여기에 함께 달겠습니다.

---

이 부분은 저도 네이밍을 어떻게 할지 고민이 됐던 부분입니다.

- text만 나타내는 cell
- 탭 가능한(expand_right 이미지가 포함된) cell
- text도 있고 탭 도 가능한 cell

이 3가지 경우에 대해서 구분할 수 있는 좀 더 적절한 네이밍을 생각해보자면..
`TextOnlySettingTableViewCell`, `TappableSettingTableViewCell`, `TappableTextSettingTableViewCell`
큰 차이는 없지만 이 정도면 괜찮을까요..? ㅎㅎ

junu0516 replied:
네 ㅋㅋ TextOnly, TextSetting 좋습니다!

junu0516 commented on Projects/DesignSystem/Sources/Enum/LifePoopAlertViewType.swift:
```diff
@@ -0,0 +1,45 @@
+
+import Foundation
+
+public enum LifePoopAlertViewType {
```
위에 커스텀 Alert 만들어주신거랑 같이 잘 봤습니다 ㅎㅎ
혹시 추후 다른 기능에서도 Alert를 보여줘야 하는 상황이면, 구현해주신 LifePoopAlertViewType에 case를 추가해서 사용하면 되는건가요??

sanghyeok-kim replied:
네 맞습니다.
LifePoopAlertViewType에 case 추가해서, 필요한 프로퍼티(title, subtitle 등)를 구현해주면 됩니다!

junu0516 replied:
감사합니다 알려주신 방식으로 저도 사용하겠습니다. :)

junu0516 commented on Projects/Features/FeatureHome/FeatureHomeCoordinatorInterface/Sources/HomeCoordinatorAction.swift:
```diff
@@ -11,4 +11,5 @@ import Foundation
 public enum HomeCoordinateAction {
     case flowDidStart
     case stoolLogButtonDidTap
+    case settingButtonDidTap
```
특정 행위를 나타내는 의미를 가진 이름이 프로젝트에서 현재 각 뷰모델의 Input 속성과 CoordinatorAction의 case에서 사용되고 있는데요,
- 주어 + 동사 : settingButtonDidTap
- 동사 + 주어 : didTapSettingButton

위와 같은 2가지 방식이 혼용되는 것 같아요. 이전에 코드 컨벤션 한 번 정리하자고 하셨는데 그 때 행위에 대해서 위의 2가지 중 적절한거 선택해서 통일하면 좋을 것 같아요

sanghyeok-kim replied:
주어 + 동사 : settingButtonDidTap
이 방법으로 통일하는게 적절할 것 같습니다.

junu0516 replied:
넵 확인 감사합니다. 저도 주어+동사로 통일할께요!
(기존에 동사 + 주어 로 되어있던 부분은 추후 수정하겠습니다.)

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingCoordinator/Sources/DefaultSettingCoordinator.swift:
```diff
@@ -0,0 +1,102 @@
+
+// MARK: - Coordinating Methods
+
+private extension DefaultSettingCoordinator {
+    func pushSettingViewController() {
+        let viewController = SettingViewController()
+        let viewModel = SettingViewModel(coordinator: self)
+        viewController.bind(viewModel: viewModel)
+        navigationController.pushViewController(viewController, animated: true)
+    }
+    
+    func pushProfileInfoViewController() {
+        let viewController = ProfileViewController()
+        let viewModel = ProfileViewModel(coordinator: self)
+        viewController.bind(viewModel: viewModel)
+        navigationController.pushViewController(viewController, animated: true)
+    }
+    
+    func pushDocumentViewController(by documentType: DocumentType) {
+        let viewController = DocumentViewController()
+        let viewModel = DocumentViewModel(documentType: documentType)
+        viewController.bind(viewModel: viewModel)
+        navigationController.pushViewController(viewController, animated: true)
+    }
+    
+    func pushFeedbackViewController() {
+        let viewController = FeedbackViewController()
+        let viewModel = FeedbackViewModel(coordinator: self)
+        viewController.bind(viewModel: viewModel)
+        navigationController.pushViewController(viewController, animated: true)
+    }
+    
+    func pushWithdrawalViewController() {
+        let viewController = WithdrawalViewController()
+        let viewModel = WithdrawalViewModel(coordinator: self)
+        viewController.bind(viewModel: viewModel)
+        navigationController.pushViewController(viewController, animated: true)
+    }
```
이건 저도 비슷하게 고쳐볼 수 있을듯한데..
혹시 지금처럼 모든 ViewController을 push 하는 로직이 동일한 형태면 제네릭 사용해서 공통함수로 빼도 괜찮을듯해요..
저는 현재까지 구현한 부분 모두 단순 push&pop 형태라서 함수 하나로 나중에 통일하려고 합니다.

sanghyeok-kim replied:
동의합니다!!
Coordinator 프로토콜의 extension으로 기본 구현을 제공해줘도 좋을 듯 하네요

junu0516 replied:
오.. extension까지 가면 더 좋네요 ㅋㅋ 모든 DefaultCoordinator이 가벼워질 것 같아요 :)

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingCoordinatorInterface/Sources/SettingCoordinatorCompletionDelegate.swift:
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+public protocol SettingCoordinatorCompletionDelegate: AnyObject {
+    func finishFlow()
```
해당 Delegate를 채택하는 쪽에서 finishFlow() 함수명만 보고 어떤 흐름이 끝난건지를 알기 힘들 수도 있을 것 같아요! 
이것도 delegate 함수명이 좀 더 구체적이면 좋을듯합니다 ㅎㅎ

물론.. 위에 코드 작성해주신대로 하나의 프로토콜 채택에 대해 각각의 extension을 가지도록 나눠주셔서 어느정도 위의 의도가 드러나긴 합니다!

sanghyeok-kim replied:
앗 이 부분은 의도적으로 단순하게 finishFlow라고 지었습니다. 
아래에 자세히 설명드릴테니 읽어보시고 제드의 생각을 말씀해주세요 ㅎㅎ

---

SettingCoordinator 입장에서 자신의 flow가 끝났을 때,
이후에 어떤 동작이 일어날지에 대해서는 자신은 모른 채로
그 책임을 부모에게 위임하는 것이 delegate pattern에 부합하다고 생각해서
일부러 finishFlow라고 메소드명을 단순하게 지었습니다.

만약 SettingCoordinator가 종료되면 SomeCoordinatorFlow가 시작된다고 가정했을 때,

메소드명을 `finishFlow`가 아니라 구체적인 `startSomeCoordinatorFlow`로 짓게 된다면
SettingCoordinator가 자신의 flow가 종료되었을 때 어떤 flow가 이어질지를 직접 알게 되는 것이고,

만약 이후에 이어질 flow가 SomeCoordinatorFlow가 아닌 다른 flow로 변경된다면
`startSomeCoordinatorFlow`이라는 메소드명도 다시 변경해줘야 하므로 변경지점이 늘어난다는 점도 고려했습니다.

junu0516 replied:
아..! 그러면 finishFlow()로 그냥 두는게 낫겠네요.
finishFlow 자체가 결국 SettingCoodinator이 포함하는 모든 화면 흐름이 종료됨을 의미하는거네요,, 이건 제가 잘못 생각했네요 ㅎㅎ

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/Document/DocumentViewController.swift:
```diff
@@ -0,0 +1,63 @@
+
+import DesignSystem
+import Utils
+
+public final class DocumentViewController: LifePoopViewController, ViewType {
+    
```
이건 혹시 DesignSystem이나 다른 공통 모듈 수준으로 사용해도 혹시 괜찮을까요?
현재 로그인화면에서 약관 보기 화면을 구현하고 있는데 사실상 코드중복인 것 같아서요,,,

sanghyeok-kim replied:
좋습니다 ㅎㅎ 그렇게 변경해서 DesignSystem에 구현해 놓을게요

DocumentViewController는 title과 textView에 들어갈 String값만 인자로 받으면 되고,
별다른 사용자 Input을 받아 처리하는 것도 없으니 ViewModel도 필요 없겠네요.
공통 컴포넌트 단위로 사용할 수 있을 것 같습니다.

junu0516 replied:
ㅎㅎ 감사합니다!

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingModels/AutoLoginSettingModel.swift:
```diff
@@ -0,0 +1,15 @@
+//  Created by 김상혁 on 2023/05/22.
+//  Copyright © 2023 LifePoop. All rights reserved.
+//
+
+import Foundation
+
+public struct AutoLoginSettingModel: SettingModel {
+    public let description: String = "자동 로그인 사용"
+    public let section: SettingListSection = .info
+    public let displayType: SettingInfoDisplayType = .switch
+}
```

1. 
AutoLoginSettingModel의 정의된 형태랑, 사용되는 부분을 확인해봤는데

- 어떤 이름의 속성을 true/false 중 하나로 업데이트할 것인지 결정
- (테이블뷰 셀로 출력되면) 스위치 버튼 요소를 포함하는 셀 형태로 출력될 것임을 암시

이 2가지 의미로 사용되는 모델인데, 혹시 추후에 기획 변경에 따라 스위치 버튼이 적용되는 옵션이 늘어날 경우를 대비해서 아얘 SwitchOptionSettingModel... 같은 식으로 좀 더 보편적 의미의 모델로 변경하는건 어떨까요!!?

지금 구현해주신 상황으로는 사실상 이름변경..만 하면 될듯합니다 😅


2. 
두번째로 section 타입 속성을 가지고 있는데, 제가 작성해주신 코드 이해한 바로는 테이블 뷰 섹션에 해당되는 SettingListSection 타입 하위로 각 SettingModel 구현체들이 존재하는 구조인 것 같아요.
만약에 이게 맞으면 SettingModel 구현체가 section을 알고 있는게, 하위가 상위를 아는 상황인듯한데.. 개인적으로는 상위 SettingListSection 수준에서 SettingModel  리스트를 알고, 각각의 SettingModel은 최대한 상위를 모르는쪽으로 가는게 좀 더 자연스러울 것 같아요!

혹시 제가 잘못 이해한거면 코멘트좀 부탁드려요 ㅎㅎ

sanghyeok-kim replied:
1 A)
아래 2번 내용을 반영하면서 수정될 예정인 부분인데요, 
SettingModel을 만들 때 '어떤 형태로 화면에 그려질 것인지'에 대한 관심사는 아예 분리할 예정이라
해당 네이밍은 이대로 사용해도 괜찮을 듯 합니다.
만약 SettingModel의 이름에 SwitchOption같은 단어가 들어가게 되면 해당 관심사가 분리되지 않으니까요.


2 A)
말씀하신대로 하위개념인 Model이 자신의 상위개념인 section을 알고 있는게 어색할 수 있겠네요.
그리고 displayType도 마찬가지로 Model이 View의 영역을 알고있는거라 어색하다고 판단이 되어서,
Model이 갖는 section과 displayType의 책임을 분리하는 작업을 진행해서 커밋 후 코멘트 남기겠습니다.

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingModels/Enum/SettingInfoDisplayType.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+public enum SettingInfoDisplayType {
+    case loginType
+    case `switch`
+    case tap
+    case textTap
+    case text
```
SettingInfoDisplayType 열거형도 마찬가지로 어느 한 '모델 타입'으로 간주하고 저는 코드를 봤는데요,
제가 작성해주신 코드 봤을 땐 테이블뷰에 출력되는 각각의 옵션을 나타내는 셀의 '뷰 출력 형태'를 결정하기 위한 성격이 강해보입니다!

혹시 제가 이해한 바가 맞다면, switch, tap, textTap, text의 4가지는 '출력 형태'의 범주에 속하지만
loginType의 경우에는 '출력 형태'라기 보다는 '구체적인 옵션 정보'에 좀 더 가까운 의미인 것 같아요.

이 부분도 괜찮으시면 로그인 정보 외에도 이미지 요소를 보여주는 셀이 늘어나는 상황에 대비해서 '출력 형태'의 의미로 바꾸는 것도 괜찮아보입니다!!

sanghyeok-kim replied:
사실 처음에 .image로 했다가 .loginType으로 변경했던 부분인데요,

cell에 출력되는 이미지가

1. '서버에 요청해서 받아온 이미지의 Data 타입을 Output으로 전달받아 UIImage로 렌더링해서 출력'
하는게 아니라,

2. '단순히 Asset에 저장되어있는 Image중 하나를 선택해서 출력하는 형태'
다 보니

1번 방법으로 이미지를 출력하는 경우와 2번 방법으로 이미지를 출력하는 방법의 ViewModel Output이 통일되지 않아서
loginType을 출력하는 별도의 타입으로 분리했습니다.

---

이 부분은 글로 설명하자니 너무 장황해질 것 같아서, 따로 회의때 논의해보면 좋을 것 같아요

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingModels/Enum/SettingListSection.swift:
```diff
@@ -0,0 +1,23 @@
+
+import Foundation
+
+public enum SettingListSection: Int {
+    case info
+    case support
+    
+    public var title: String {
+        switch self {
+        case .info:
+            return "내 정보"
+        case .support:
+            return "고객 지원"
+        }
+    }
+}
```
위에 SettingModel 관련해서 남긴 코멘트랑 연결되는 부분인데,
섹션에 해당하는 SettingListSection 안에서 SettingModel 배열을 들고있는 형태도 괜찮아보입니다!

junu0516 commented on Projects/Utils/Sources/Enum/LoginType.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+public enum LoginType {
+    case apple
+    case kakao
+}
```
해당 타입은 인증 과정에서도 사용될 수 있을 것 같아요. 혹시 CoreEntity 모듈에 두는건 어떤가요?

sanghyeok-kim replied:
이 부분도 우선은 임시적으로 Utils에 단순 열거형으로 만들어놓은거라,
나중에 로그인 기능에 대한 UseCase를 구현하면서 계속 수정해나가야 할 것 같습니다.

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/Withdrawal/WithdrawalViewController.swift:
```diff
@@ -0,0 +1,152 @@
+        return label
+    }()
+    
+    private let reasonButtons = WithdrawReason.allCases.map {
+        SelectableTextRadioButton(index: $0.index, title: $0.title)
+    }
+    
+    private lazy var reasonSelectStackView: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: reasonButtons)
+        stackView.axis = .vertical
+        stackView.spacing = 16
+        stackView.alignment = .leading
+        stackView.distribution = .equalSpacing
+        return stackView
+    }()
```
저는 비슷한 유형의 뷰를 컬렉션뷰(수평 스크롤)로 구현했는데, CaseIterable을 통해서 복수의 버튼 요소 + 스택뷰로도 구현할 수 있네요! 덕분에 다른 방식 잘 보고 배웠어요 😁

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/Withdrawal/WithdrawalViewModel.swift:
```diff
@@ -0,0 +1,87 @@
+    
+    public struct Input {
+        let viewDidLoad = PublishRelay<Void>()
+        let reasonDidSelectAt = PublishRelay<Int>()
+        let withdrawButtonDidTap = PublishRelay<Void>()
+        let withdrawCancelButtonDidTap = PublishRelay<Void>()
+        let withdrawConfirmButtonDidTap = PublishRelay<Void>()
```
탈퇴 사유 선택을 위해, 뷰에서 스택뷰 안에 여러 개의 버튼을 배치시키고 각 버튼 터치에 대한 input 속성들이 여기에 정의된거죠?
이 부분도 잘 봤습니다 ㅎㅎ

감상을 하나 남겨보자면..
스택뷰 + 버튼들 조합이 컬렉션뷰(혹은 테이블뷰)보다 뷰 차원에서 가벼워서 좋은 것 같고,
뷰모델 차원에서는 각 버튼들에 대한 input을 모두 열거해야 하는 점이 차이인 것 같아요!

sanghyeok-kim replied:
스택뷰도 사실 렌더링이 무거운 작업이라서 성능상으로 가벼운건 모르겠네요 😅 구현상 편리한건 확실합니다 ㅎㅎ

둘 다 각각의 장단점이 있는 것 같은데,
다만 이처럼 특정 값만 다른 동일한 형태의 뷰가 연달아 나타나게 뷰를 출력해야 할 때
TableView나 CollectionView처럼 선택지가 많아서 스크롤이 필요한 경우(따라서 성능상 reuse가 필요한 경우)가 아니라면
이를 구현하는 Best Practice가 CaseIterable 열거형 + StackView라고 생각해서 그렇게 구현해보았습니다

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingViewModel.swift:
```diff
@@ -0,0 +1,200 @@
+            .bind(to: state.version)
+            .disposed(by: disposeBag)
+        
+        input.viewDidLoad
+            .withUnretained(self)
+            .map { `self`, _ in
+                SettingType.allCases.map { self.generateSettingCellViewModel(by: $0) }
+            }
+            .bind(to: output.settingCellViewModels)
+            .disposed(by: disposeBag)
```
위에서 얘기했던 것들과 연결되는데, 여기서 최종적으로 settingCellViewModels를 포함하는 section 모델을 가지고 전체 테이블뷰 구조를 출력하면 직관적일듯 해요

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingViewModel.swift:
```diff
@@ -0,0 +1,200 @@
+            .bind(to: output.settingCellViewModels)
+            .disposed(by: disposeBag)
+        
+        input.viewDidLoad
+            .map { SettingTableFooterViewModel() }
+            .do { [weak self] in
+                self?.bind(settingTableFooterViewModel: $0)
+            }
+            .bind(to: output.footerViewModel)
+            .disposed(by: disposeBag)
```
헤더, 푸터 모델도 섹션에 포함시켜셔, 헤더, 푸터, 셀 3가지 출력을 섹션 데이터 하나로 일원화할 수 있을듯합니다.

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingViewModel.swift:
```diff
@@ -0,0 +1,200 @@
+                model: settingType.model,
+                tapAction: input.feedbackDidTap
+            )
+        }
```
현재 구조로는, 기획단에서 설정화면 옵션 추가/수정될 때마다 `tapAction` 으로 넘겨주는 input 속성의 추가 및 삭제가 뷰모델 안에서 매번 발생할 것 같은데.. 기왕 tapAction 파라미터로 각 셀 뷰모델이 취하는 '구체적인 설정 행위'를 정의해주신 김에 tapAction이 받는 타입도 publishRelay(Void)가 아닌 '구체적인 설정 행위'의 의미를 포함하는 모델(혹은 이를 감싸는 Relay)로 정의해서, SettingCellViewModel이 각각 들고있는 형태도 괜찮을듯 해요.

이렇게 할 경우에는 SettingViewModel 수준에서 모든 옵션 행위를 일일히 알지 않아도, SettingViewModel이 현재 가지고 있는 SettingCellViewModel이 해당 셀/셀뷰모델이 나타내는 '옵션 뷰 출력'에 대한 정보 뿐만 하니라, 해당 뷰에 연결되는 '설정 행위'도 같이 알기 때문에 각 셀 뷰모델들의 존재의미가 좀 더 분명해지지 않을까 생각듭니다!

현재 구조를 쭉 봤을 때, SettingViewModel 안에서 테이블뷰의 각 셀에 해당되는 '옵션' 하나에 필요한, '뷰 출력 정보', '구체적인 설정 행위' 2가지 정보가 한군데 있지 않고 서로 흩어져있는 형태여서 이걸 한군데 있도록 하면 어떨까 싶어 생각한 부분입니다!

예를 들어, 스위치 버튼을 사용하는 셀이면 해당 셀뷰모델은 타입 속성이 .switch 이런 식으로 되있을 텐데, 그 안에 ()->Bool 타입 클로저를 포함시켜서 해당 셀뷰모델이 스위치 버튼을 출력하는 뷰이면서, 내부적으로 Bool값을 사용하는 '설정 행위'를 가지는 것임을 같이 나타낼 수 있겠다 생각했어요.

sanghyeok-kim replied:
> 예를 들어, 스위치 버튼을 사용하는 셀이면 해당 셀뷰모델은 타입 속성이 .switch 이런 식으로 되있을 텐데,
그 안에 ()->Bool 타입 클로저를 포함시켜서

저도 구현할 때 고민했던 점인데, 우선 열거형 안에 연관타입을 넣게 되면 CaseIterable 채택에 제한이 있더라구요.

그렇다고 SettingModel을 채택한 구현 타입마다 각각의 Action에 해당하는 클로저를 받도록 하자니,
해당 SettingModel 프로코콜을 사용하는 쪽에서 다시 구체타입으로 다운 캐스팅해줘야 하는 문제도 있었구요.

이 부분도 글로 설명하긴 힘들 것 같아서 회의에서 같이 코드 보면서 자세히 논의해보면 좋을 것 같습니다!

junu0516 replied:
음.. 저는 열거형의 연관타입이 아니라 클래스 형태로 된 각 셀 뷰모델 혹은 셀 뷰모델 생성을 위해 넘겨주는 SettingModel 구현체가 클로저 혹은 클로저를 감싸는 모델을 가지면 어떨까 생각했어요.

제가 간단하게 생각한바로는 아래와 같이 클로저를 포함하는 모델의 인터페이스를 정의하고,
```swift
protocol DevModeAction {
    associatedtype AttributeType: DevModeAttribute
    associatedtype ActionResult
    associatedtype Action = (@escaping (Any...) -> Void) -> ActionResult

    var attribute: AttributeType { get }
    var perform: Action? { get }
}
```

구현체쪽에서는 원하는 모든 형식의 클로저로 구현할 수 있도록 하는 식으로도 검토 가능할 것 같아요.
```swift
// 스위치 액션의 경우
struct DevModeSwitchAction: DevModeAction {
    typealias ActionResult = Bool
    typealias Action = (SwitchTypeAttribute) -> Bool

    let attribute: SwitchTypeAttribute
    let perform: Action?
}
// 텍스트필드 액션의 경우
struct DevModeTextFieldAction: DevModeAction {
    typealias ActionResult = String
    typealias Action = (ActionParameter) -> String

    struct ActionParameter {
        let value: String
    }

    let attribute: TextFieldTypeAttribute
    let perform: Action?
}
```

골자는, 어느 한 옵션에 대해 '옵션의 정의'와 '옵션의 행위'가 뷰모델 안에서 서로 흩어져 있는 구조를, 정의와 행위가 한 군데 모여있는 쪽으로 최대한 수정하면 좋겠다 입니다!

sanghyeok-kim replied:
말씀하신대로 옵션의 '정의'와 '행위'를 모두 Model 안에 합쳐서 구현하지 않는 이유는
MVVM 관점에서 '정의'에 대한 책임은 Model이 갖고, '행위'에 대한 책임은 ViewModel이 갖는게 적절하다고 생각해서 그랬습니다.
따라서 Model이 행위와 정의를 모두 갖는 것 보다는 지금처럼 ViewModel과 Model이 각각 분리해서 갖고있는게 더 낫지 않을까요?

Setting~CellViewModel들은 자신이 가져야할 정의(어떤 세팅 타입인지)는 Model을 통해 전달받고,
행위(View Input과 관련된 Action)에 대한 처리는 상위 ViewModel에게 전달하는게 맞다고 생각해서
저렇게 각각 분리해서 전달했던거였어요.

다만 전달해야할 Action이 많아지게되면 별도의 타입으로 분리해서 전달해야할 필요는 있어 보이네요

더 자세한 내용은 오늘 회의때 논의해보면 좋겠습니다 ㅎㅎ

junu0516 replied:
아하 ㅎㅎㅎ '정의'와 '행위'를 각각 어디 두냐에 따라 각기 다른 관점이군요 결국 ㅋㅋ
뷰모델이 행위에 대한 책임을 갖는거면, 전체 뷰모델보다는 개별 셀 뷰모델이 가지는게 아무래도 전체 뷰모델을 더 가볍게 하지 않을까 싶어요 그러면!

우선은 현재 구조가 담당자가 코드 관리 및 수정을 감당하기 제일 좋은 구조인듯 하고, 제 3자가 봐도 흐름을 파악하는데 문제 없으니까 이대로 두면 될 것 같아요! 
덕분에 저도 새로운 관점 하나 배워갑니다~!

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingViewModel.swift:
```diff
@@ -0,0 +1,200 @@
+        
+        input.feedbackDidTap
+            .bind { coordinator?.coordinate(by: .sendFeedbackDidTap) }
+            .disposed(by: disposeBag)
```
저는 bind(onNext:)를 썼는데, 오히려 비효율적으로 작성하고 있었네요 🤣 이 부분 보고 저도 고쳐보겠습니다.

junu0516 reviewed (changes requested):
이전보다 작성해주신 코드량이 상대적으로 많아서 돌려보고, 구조 전반적으로 짜주신거 살펴보느라 좀 걸렸어요 😅
테이블 뷰 구조가 다른 화면들에 비해 좀 복잡해서 이것저것 고민 많으셨을텐데 수고 많으셨어요~! 보면서 저도 이것저것 배워갑니다.
 
보면서 좀 궁금한 부분도 있고, 설정 옵션을 여러 모델로 구현해주신 부분에 대해서 개인적으로 여러 가지 생각도 들어서.. 우선은 RequestChanges로 리뷰 제출했습니다.!

시간 괜찮으실 때 메이스 일정상 답변하시기 편한 부분만 코멘트좀 부탁드려요 ㅎㅎ
코멘트 달아주시면, 이후에 추가 작업 올리실 부분 없으면 Approve하고 바로 머지하면 될듯합니다 ㅎㅎ

junu0516 commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingViewController.swift:
```diff
@@ -0,0 +1,137 @@
+import RxSwift
+import SnapKit
+
+import DesignSystem
+import Utils
+
+public final class SettingViewController: LifePoopViewController, ViewType {
+    
+    private let settingTableViewDelegate = SettingTableViewDelegate()
```
@sanghyeok-kim  이 부분 물어본달걸 깜빡해서.. 추가로 하나만 더 질문 남길께요
LifePoopViewController과 ViewType을 하나로 합칠 수는 없을까요?
다른 뷰컨트롤러도 LifePoopViewController 상속처리해달라고 하셔서, 이거 보다가 매번 모든 뷰컨트롤러에 ViewType과 LifePoopViewController 2개의 타입을 동시 채택할거면, 2개를 합치는게 낫지 않을까 생각이 들었어요

sanghyeok-kim replied:
그러려면 우선 DesignSystem 모듈에 Utils 모듈 의존성이 추가되어야하고(ViewType을 알기 위해),

LifePoopViewController에 ViewType을 채택하더라도
LifePoopViewController의 ViewModel 타입을 특정할 수 없기 때문에 구현할 수 없을 것 같습니다
(ViewType을 채택한 클래스는 associatedtype인 ViewModel이 어떤 타입인지 선언해야하므로)
<img width="562" alt="CleanShot 2023-05-23 at 10 16 04@2x" src="https://github.com/LifePoop/LifePoop_iOS/assets/57667738/0c9cfd5a-4541-4e98-bac6-ff4093acfe3c">

LifePoopViewController는 지금처럼 디자인 시스템 본연의 역할-화면에 그려지는 것-만 수행하도록 하고,
그걸 서브클래싱한 뷰컨에서 ViewModel이 필요할 경우에만 ViewType을 채택해서 별도의 바인딩 로직을 구현해주는게 나을 듯 합니다.

예외적으로 LifePoopViewController를 서브클래싱했지만 ViewModel이 필요없는 뷰컨도 있을테니까요
(위에 코멘트하신 [DocumentViewController](https://github.com/LifePoop/LifePoop_iOS/pull/42#discussion_r1200332523)가 그 경우에 해당합니다)

junu0516 replied:
그런 이유면 결국 LifePoopViewController, ViewType 2가지가 서로 각기 다른 존재의미를 지니네요

- LifePoopViewController : UI 속성 설정, 레이아웃 설정을 위한 인터페이스를 정의하기 위함
- ViewType: 뷰모델과 바인딩되는 '뷰'의 특성을 나타내기 위함

이렇게 정리되는듯 한데, LifePoopViewController의 경우에는 사실상 UI 및 레이아웃 설정을 위함 '함수명 통일'과 일부 반복되는 배경색상 및 네비게이션 아이템 설정 2가지 역할을 하는데 차라리 이걸 각각의 인터페이스로 쪼개는게 좀 더 직관적일 수도 있을듯해요.

위의 내용 다 차치하더라도, 메이스가 의도하신 `'모든 뷰컨트롤러가 LifePoopViewController을 상속받아야 함'` 의 의미가 좀 더 분명하고, 추후 제 3자가 굳이 말로 전달받지 않아도 뷰컨트롤러는 LifePoopViewController을 상속받아야 함을 자연스럽게 알 수 있었으면 좋겠어요

sanghyeok-kim replied:
> UI 및 레이아웃 설정을 위함 '함수명 통일'과 일부 반복되는 배경색상 및 네비게이션 아이템 설정 2가지 역할을 하는데 차라리 이걸 각각의 인터페이스로 쪼개는게

'배경색상 및 네비게이션 아이템 설정' 역할이 `configureUI`라는 메소드를 통해 실행되고, 
이후 서브클래싱한 클래스들이 해당 메소드를 오버라이딩해서 쓰는 과정에서 '함수명 통일' 역할도 동시에 연계되는거라서
2가지 역할이 서로 분리되어있다기보다는 사실상 합쳐져서 하나의 역할을 하고 있다고 보시면 될 거 같습니다.

---

의도가 명확하게 드러나지 않는건 네이밍의 문제인 것 같기도 해서
LifePoopViewController 말고 BaseViewController라고 정의하는게 더 나을까요?

junu0516 replied:
아 ㅎㅎㅎ 저는 이름은 괜찮은거 같아요. LifePooButton처럼 그냥 공통으로 쓰는 타입 중 하나로 충분히 의미전달 되는 것 같습니다. 다만, ViewType이라는게 또 존재하다보니까 2가지의 존재 의미를 좀 더 분명히하고자 여쭤봤어요 ㅋㅋ
다시 생각해보니까, 제가 무의식적으로 모든 뷰컨트롤러가 ViewType, LifePooViewController을 다 채택 및 상속받아야 한다고 생각했는데 말씀하신대로 어떤 상황에서는 ViewType이 아닌 뷰컨트롤러로 존재해야 하는 경우도 있으니까 이정도로 우선 코드 작성 의도 전달된걸로 두면 될듯해요 ㅎㅎ
우선 만들어주신 의도는 충분히 이해해서 저도 추후에는 의도하신 바에 맞게 작업하면 될 것 같습니다..!

sanghyeok-kim committed:
#20 refactor: SettingModel의 section, displayType 관심사 분리

sanghyeok-kim committed:
#20 refactor: Factory를 구현하여 SettingType이 직접 자신의 Model을 생성하는 책임을 분리

sanghyeok-kim committed:
#20 refactor: SettingModel로부터 분리된 Section, DisplayType에 대한 책임을 갖는 별도의 Mapper 구현

sanghyeok-kim committed:
#20 refactor: 변경된 SettingModel에 맞춰 SettingViewModel, SettingTableViewDataSource의 코드 수정

sanghyeok-kim committed:
#20 refactor: SettingViewCell 구성 및 뷰모델 바인딩 작업을 별도의 CellFactory로 분리

sanghyeok-kim committed:
#20 refactor: SettingSectionMapper를 변경된 이름 SettingModelSectionMapper로 수정

sanghyeok-kim committed:
#20 refactor: DocumentViewController를 DesignSystem 컴포넌트로 분리, 뷰모델 의존성 제거

sanghyeok-kim committed:
#20 refactor: SettingTapTableViewCell, SettingTextTapTableViewCell을 SettingTapActionTableViewCell로 통합

- Tap Action을 나타내는 expand_right 이미지와 additionalTextLabel을 기본으로 갖고, additionalTextLabel의 text는 옵셔널로 나타냄

sanghyeok-kim commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingTableView/ViewCell/SettingTapActionTableViewCell.swift:
```diff
@@ -0,0 +1,90 @@
+import DesignSystem
+import Utils
+
+public class SettingTapActionTableViewCell: BaseSettingTableViewCell, ViewType {
+    
+    public lazy var addtionalTextLabel: UILabel = {
+        let label = UILabel()
+        label.font = .systemFont(ofSize: 16)
+        return label
+    }()
+    
+    public lazy var expandImageView: UIImageView = {
```
기존에 tap, textTap 으로 나눠져 있던 Cell의 형태를 tap 하나로 통일했습니다.
다만 tap의 경우 additionalText를 옵셔널로 받아서, text가 있다면 expand_right 이미지 왼쪽에 나타내는 식으로 했습니다.

따라서 DesignSystem에는 BaseSettingTableViewCell만 남기고 나머지는 삭제했고,

DisplayType도 tap이냐 textTap이냐에 따라 각각 다른 viewCell과 cellViewModel 타입을 캐스팅해줄 필요 없이
일괄적으로 SettingTapActionTableViewCell, SettingTapActionCellViewModel로 통일했습니다.

sanghyeok-kim commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingModels/Enum/SettingInfoDisplayType.swift:
```diff
@@ -12,6 +12,5 @@ public enum SettingInfoDisplayType {
     case `switch`
     case tap
-    case textTap
     case text
```
[코멘트 내용](https://github.com/LifePoop/LifePoop_iOS/pull/42#discussion_r1201631912)

sanghyeok-kim commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingTableView/CellViewModel/SettingTapActionCellViewModel SettingTapActionCellViewModel.swift:
```diff
@@ -22,16 +22,18 @@ public final class SettingTapCellViewModel: SettingCellViewModel {
     
-    public init(model: SettingModel, tapAction: PublishRelay<Void>) {
+    public init(model: SettingModel, tapAction: PublishRelay<Void>, additionalText: BehaviorRelay<String>? = nil) {
         self.model = model
```
[코멘트 내용](https://github.com/LifePoop/LifePoop_iOS/pull/42#discussion_r1201631912)

sanghyeok-kim commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingModels/Enum/SettingType.swift:
```diff
@@ -16,23 +16,4 @@ public enum SettingType: CaseIterable {
-        case .sendFeedback:
-            return SendFeedbackSettingModel()
-        }
-    }
```
SettingType 열거형이 Model을 직접 생성하는 책임까지 갖지 않도록 Factory로 분리했습니다

sanghyeok-kim commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingModels/Interface/SettingModel.swift:
```diff
@@ -10,6 +10,5 @@ import Foundation
-    var section: SettingListSection { get }
-    var displayType: SettingInfoDisplayType { get }
+    var type: SettingType { get }
 }
```
SettingModel이 갖는 section과 displayType에 대한 관심사를 제거하고, 대신 SettingType을 설정하였습니다.
이후 SettingType에 맞는 section과 displayType을 맵핑해주는 책임을 갖는 Mapper 객체를 별도로 구현하였습니다.

sanghyeok-kim commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingModels/Factory/SettingModelDisplayTypeMapper.swift:
```diff
@@ -0,0 +1,26 @@
+        switch model.type {
+        case .loginType:
+            return .loginType
+        case .profile:
```
[코멘트 내용](https://github.com/LifePoop/LifePoop_iOS/pull/42#discussion_r1201644337)

sanghyeok-kim commented on Projects/Features/FeatureSetting/FeatureSettingPresentation/Sources/SettingModels/Factory/SettingModelSectionMapper.swift:
```diff
@@ -0,0 +1,20 @@
+
+import Foundation
+
+struct SettingModelSectionMapper {
+    static func mapSection(for model: SettingModel) -> SettingListSection {
+        switch model.type {
+        case .loginType, .profile, .autoLogin, .version:
+            return .info
```
[코멘트 내용](https://github.com/LifePoop/LifePoop_iOS/pull/42#discussion_r1201644337)

sanghyeok-kim committed:
#20 fix: DocumentViewController의 textView가 editing되는 문제 수정

sanghyeok-kim commented:
- SettingModel이 Section과 DisplayType을 알고 있는 문제를 팩토리 패턴 적용해서 책임 분리
- .tap과 .textTap 두 종류로 나눠져 있던 Cell의 DisplayType을 하나로 통일
- DocumentViewController를 DesignSystem 모듈의 공통 컴포넌트로 분리
이렇게 리팩토링 진행했고, 추가적인 논의 사항은 회의때 이야기하면 될 것 같습니다!
